File: compiler/src/error_types.py
"""Error type constants for machine-readable compile error categorization."""

class ErrorType:
    """Error type constants for compile errors."""
    pass

# Auto-generate error type constants
_ERROR_TYPES = [
    # Syntax and structure errors
    "EXPECTED_DO_AFTER",           # Expected 'do' after 'for', 'while', etc.
    "EXPECTED_THEN_AFTER",         # Expected 'then' after 'if'
    "SYNTAX_ERROR",                # General syntax errors
    "INVALID_MACRO",               # Unknown macro
    "INVALID_INDENTATION",         # Invalid indentation (spaces instead of tabs)
    
    # Type checking errors
    "TYPE_MISMATCH",               # Type mismatch in field/argument
    "FIELD_TYPE_MISMATCH",         # Field demands specific type but given another
    "ARGUMENT_TYPE_MISMATCH",      # Argument demands specific type but given another
    "MISSING_TYPE",                # Field demands type but given None
    
    # Structure/argument errors  
    "WRONG_ARG_COUNT",             # Wrong number of arguments
    "MISSING_CHILD",               # Missing required child node
    "MISSING_BLOCK",               # Missing required block (do, then, etc.)
    "INVALID_STRUCTURE",           # Invalid node structure
    
    # Scope and context errors
    "MISSING_CONTEXT",             # Required context missing (e.g., param outside fn)
    "INVALID_PLACEMENT",           # Macro in wrong place
    
    # String literal errors
    "UNTERMINATED_STRING",         # String not properly delimited
    
    # General errors
    "ASSERTION_FAILED",            # General assertion failure
    "UNKNOWN_ERROR",               # Uncategorized error
]

# Set each error type as a class attribute with the same string value
for error_type in _ERROR_TYPES:
    setattr(ErrorType, error_type, error_type)

File: compiler/src_ts/utils/error_types.ts
// utils/error_types.ts

// Add/remove names here; everything else updates automatically.
export const ERROR_TYPES = [
  "EXPECTED_DO_AFTER",
  "EXPECTED_THEN_AFTER",
  "SYNTAX_ERROR",
  "INVALID_MACRO",
  "INVALID_INDENTATION",

  "TYPE_MISMATCH",
  "FIELD_TYPE_MISMATCH",
  "ARGUMENT_TYPE_MISMATCH",
  "MISSING_TYPE",
  "NO_SUCH_LOCAL",
  "NO_MATCHING_OVERLOAD",
  "AMBIGUOUS_OVERLOAD",

  "WRONG_ARG_COUNT",
  "MISSING_CHILD",
  "MISSING_BLOCK",
  "INVALID_STRUCTURE",

  "MISSING_CONTEXT",
  "INVALID_PLACEMENT",

  "UNTERMINATED_STRING",

  "INVALID_INT",
  "INVALID_FLOAT",

  "ASSERTION_FAILED",
  "UNKNOWN_ERROR",
] as const;

// TypeScript literal-union of all names.
export type ErrorType = typeof ERROR_TYPES[number];

// Simple namespace-like lookup (Crystal-style `ErrorType::FOO`)
export const ErrorType: Record<ErrorType, ErrorType> = Object.fromEntries(
  ERROR_TYPES.map((e) => [e, e]),
) as Record<ErrorType, ErrorType>;
